crates/modules/bank/src/lib.rs

```diff
@@ -0,0 +1,560 @@
+            ctx: &mut Context,
+            denom: &str,
+            amount: u128,
+            evt: EventBus<EventBurn<'_>>,
+        ) -> Result<()>;
+    }
+
+    /// The send hook is called when a coin is sent to another account.
+    #[handler_api]
+    #[automock]
+    pub trait SendHook {
+        /// Called when a coin is sent to another account.
+        fn on_send<'a>(
+            &self,
+            ctx: &mut Context<'a>,
+            from: AccountID,
+            to: AccountID,
+            denom: &str,
+            amount: u128,
+        ) -> Result<()>;
+    }
+
+    /// The burn hook is called when a coin is burned from an account.
+    #[handler_api]
+    #[automock]
+    pub trait BurnHook {
+        /// Called when a coin is burned from an account.
+        fn on_burn<'a>(
+            &self,
+            ctx: &mut Context<'a>,
+            from: AccountID,
+            denom: &str,
+            amount: u128,
+        ) -> Result<()>;
+    }
+
+    /// The receive hook is called when a coin is received by an account.
+    #[handler_api]
+    #[automock]
+    pub trait ReceiveHook {
+        /// Called when a coin is received by an account.
+        fn on_receive<'a>(
+            &self,
+            ctx: &mut Context<'a>,
+            from: AccountID,
+            to: AccountID,
+            denom: &str,
+            amount: u128,
+        ) -> Result<()>;
+    }
+
+    /// An event emitted when a coin is sent to another account.
+    #[derive(SchemaValue, Clone, Default)]
+    #[non_exhaustive]
+    pub struct EventSend<'a> {
+        /// The account that sent the coin.
+        pub from: AccountID,
+        /// The account that received the coin.
+        pub to: AccountID,
+        /// The coin that was sent.
+        pub coin: Coin<'a>,
+    }
+
+    /// An event emitted when a coin is minted for an account.
+    #[derive(SchemaValue, Clone, Default)]
+    #[non_exhaustive]
+    pub struct EventMint<'a> {
+        /// The account that minted the coin.
+        pub to: AccountID,
+        /// The coin that was minted.
+        pub coin: Coin<'a>,
+    }
+
+    /// An event emitted when a coin is burned from an account.
+    #[derive(SchemaValue, Default, Clone)]
+    #[non_exhaustive]
+    pub struct EventBurn<'a> {
+        /// The account that burned the coin.
+        pub from: AccountID,
+        /// The coin that was burned.
+        pub coin: Coin<'a>,
+    }
+
+    /// The bank module used for transfering tokens between accounts.
+    impl Bank {
+        /// Called when the bank module is created.
+        #[on_create]
+        pub fn create(&self, ctx: &mut Context) -> Result<()> {
+            self.super_admin.set(ctx, ctx.caller())?;
+            Ok(())
+        }
+
+        /// Create a new denom.
+        #[publish]
+        pub fn create_denom(&self, ctx: &mut Context, denom: &str, admin: AccountID) -> Result<()> {
+            ensure!(self.super_admin.get(ctx)? == ctx.caller(), "not authorized");
```

Comment:
only the super admin can create denoms?

this is less token factoryish where everybody can and denom is constructed as "creator/denom"

Reply:
true, ill create a pr to fix this